### PR TITLE
yubihsm-shell: fix linking to pcsclite.h

### PR DESCRIPTION
**Pull request by jnsgruk**
## Description of changes

The build of this package failed for me, because one of the dependencies cannot link `pcsclite.h`. This seems to fix the problem for me, though there may be more elegant ways to fix?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/security/yubihsm-shell/default.nix [hunk 1]
```diff
@@ -48,12 +48,17 @@ stdenv.mkDerivation rec {
...
+    pcsclite.dev
   ] ++ lib.optionals stdenv.isDarwin [
     darwin.apple_sdk.frameworks.PCSC
     libiconv
   ];
 
+  buildPhase = lib.optional stdenv.isLinux ''
+    NIX_CFLAGS_COMPILE="$(pkg-config --cflags libpcsclite) $NIX_CFLAGS_COMPILE"
+    runHook buildPhase
+  '';
```
Comment (drupol) [inline pkgs/tools/security/yubihsm-shell/default.nix:60]:
```suggestion
  buildPhase = lib.optionalString stdenv.isLinux ''
    runHook preBuild
    NIX_CFLAGS_COMPILE="$(pkg-config --cflags libpcsclite) $NIX_CFLAGS_COMPILE"
    runHook postBuild
  '';
```

Comment (drupol) [reply pkgs/tools/security/yubihsm-shell/default.nix:60]: Or maybe we could simply use a `preBuild` step ?

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/tools/security/yubihsm-shell/default.nix (hunk 1 above)
Comment (jnsgruk) [reply pkgs/tools/security/yubihsm-shell/default.nix]: Good call - adjusted!

File: pkgs/tools/security/yubihsm-shell/default.nix [hunk 2]
```diff
@@ -48,12 +48,16 @@ stdenv.mkDerivation rec {
...
-    pcsclite
+    pcsclite.dev
   ] ++ lib.optionals stdenv.isDarwin [
     darwin.apple_sdk.frameworks.PCSC
     libiconv
   ];
 
+  preBuild = lib.optional stdenv.isLinux ''
+    NIX_CFLAGS_COMPILE="$(pkg-config --cflags libpcsclite) $NIX_CFLAGS_COMPILE"
+  '';
```
Comment (drupol) [inline pkgs/tools/security/yubihsm-shell/default.nix:59]: Is `preBuild` meant to be a string or a list?

If it's a list, I would prefer to use `lib.optionals`.
If it's a string, then `lib.optionalString` should be use.

Comment (jnsgruk) [reply pkgs/tools/security/yubihsm-shell/default.nix]: changed to `lib.optionalString`

**Review by drupol — APPROVED** [review]

**Comment by jnsgruk** [comment]
Result of `nixpkgs-review pr 334216` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>step-kms-plugin</li>
    <li>yubihsm-shell</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
If it is failing in one of the dependencies, we should rather propagate pcslite there to fix this universally.

**Comment by jnsgruk** [comment]
> If it is failing in one of the dependencies, we should rather propagate pcslite there to fix this universally.

I *think* it's failing in the yubihsm-shell build itself, unless I'm mistaken.

**Comment by eclairevoyant** [comment]
Based on https://cache.nixos.org/log/526wj9mpch6sp0gwhfgxjhnx6bihh3h5-yubihsm-shell-2.5.0.drv:

```
[ 23%] Building C object ykhsmauth/CMakeFiles/ykhsmauth.dir/error.c.o
[ 23%] Building C object ykhsmauth/CMakeFiles/ykhsmauth.dir/ykhsmauth.c.o
In file included from /build/source/ykhsmauth/internal.h:26,
                 from /build/source/ykhsmauth/ykhsmauth.c:22:
/nix/store/rr2bj6yaa3xn23bakcrhsz756vawyzb0-pcsclite-2.2.3-dev/include/PCSC/winscard.h:41:10: fatal error: pcsclite.h: No such file or directory
   41 | #include <pcsclite.h>
      |          ^~~~~~~~~~~~
compilation terminated.
```

I'm reading this as a header file in pcsclite.dev failing to find another header in that package

**Comment by jnsgruk** [comment]
> I'm reading this as a header file in pcsclite.dev failing to find another header in that package

Interesting, yet I can see them side-by-side in the nix store, any idea how that might get resolved? 

```
❯ ls -la /nix/store/rr2bj6yaa3xn23bakcrhsz756vawyzb0-pcsclite-2.2.3-dev/include/PCSC
.r--r--r-- 5.4k root root  1 Jan  1970 debuglog.h
.r--r--r--  31k root root  1 Jan  1970 ifdhandler.h
.r--r--r--  16k root root  1 Jan  1970 pcsclite.h
.r--r--r--  16k root root  1 Jan  1970 reader.h
.r--r--r-- 4.2k root root  1 Jan  1970 winscard.h
.r--r--r-- 3.1k root root  1 Jan  1970 wintypes.h
```

**Comment by jnsgruk** [comment]
It also looks like the pkg-config file has been installed in the `dev` output:

```
❯ tree
.
├── bin
│  └── pcsc-spy
├── include
│  └── PCSC
│     ├── debuglog.h
│     ├── ifdhandler.h
│     ├── pcsclite.h
│     ├── reader.h
│     ├── winscard.h
│     └── wintypes.h
├── lib
│  └── pkgconfig
│     └── libpcsclite.pc
└── nix-support
   └── propagated-build-inputs
```

**Comment by jnsgruk** [comment]
Looking at https://github.com/LudovicRousseau/PCSC/issues/89 and https://discourse.nixos.org/t/another-missing-header-question/7119 makes me think this solution is likely a reasonable one?

**Comment by jnsgruk** [comment]
Okay, based on the following from https://wiki.nixos.org/wiki/C:

> However, while the $out/include folder will be included by default, it may sometimes not be enough when the lib puts the header in a subfolder (for instance, gtk2 and gtk3 uses subdirectories like $out/include/gtk-2.0 instead to avoid conflict between versions). To deal with this kind of libraries, one can use `pkg-config`: the idea is simply to add `pkg-config` in the nativeBuildInputs, and then to start the buildPhase with: <code snippet that looks like this PR>

I think I'll merge this. Happy to adjust later if required.

**Comment by jnsgruk** [comment]
I've run the build (without my patch) with the CMake flags `--trace` and `-DCMAKE_VERBOSE_MAKEFILE=ON`, which resulted in the following build log:

https://gist.github.com/jnsgruk/58056b865692638271683b8f93eb2e42

(cc: @emilylange)

**Comment by emilazy** [comment]
Seems like the failing file is just not getting the `pkg-config` stuff passed down at all:

```
cd /build/source/build/ykhsmauth && /nix/store/62zpnw69ylcfhcpy1di8152zlzmbls91-gcc-wrapper-13.3.0/bin/gcc -DDEFAULT_CONNECTOR_URL=\"http://localhost:12345\" -DHAVE_EXPLICIT_BZERO -DOPENSSL_API_COMPAT=0x10000000L -DYKHSMAUTH_ENABLED=\"1\" -D_FORTIFY_SOURCE=2  -flto -Wno-missing-braces -Wno-missing-field-initializers -O3 -DNDEBUG -std=gnu11 -fPIC -DSTATIC -Wall -Wextra -Werror -Wformat -Wformat-nonliteral -Wformat-security -Wshadow -Wcast-qual -Wmissing-prototypes -Wbad-function-cast -pedantic -pedantic-errors -fpie -fpic -O2 -fstack-protector-all -Wno-implicit-fallthrough -MD -MT ykhsmauth/CMakeFiles/ykhsmauth_static.dir/ykhsmauth.c.o -MF CMakeFiles/ykhsmauth_static.dir/ykhsmauth.c.o.d -o CMakeFiles/ykhsmauth_static.dir/ykhsmauth.c.o -c /build/source/ykhsmauth/ykhsmauth.c
```

I suspect this to be an upstream build system bug. They do seem to [set the include path](https://github.com/Yubico/yubihsm-shell/blob/4d6142e919003894b16d278ff52b424882ee5300/ykhsmauth/CMakeLists.txt#L28), though. Maybe our `libpcsclite.pc` is messed up somehow?

**Comment by jnsgruk** [comment]
> Maybe our `libpcsclite.pc` is messed up somehow?

The contents of our pkgconfig file is below, I wonder if that libdir line should really end `lib/PCSC` in order for it to work - is that something we could just patch in the postInstall?

```
prefix=/nix/store/0v1qyxisn7hhf63wg3c6kpc7dwhcn90w-pcsclite-2.2.3
includedir=/nix/store/rr2bj6yaa3xn23bakcrhsz756vawyzb0-pcsclite-2.2.3-dev/include
libdir=/nix/store/9pb5dql4fnbdrq8wsjan0la13b2hwgi0-pcsclite-2.2.3-lib/lib

usbdropdir=/var/lib/pcsc/drivers
serialconfdir=/etc/reader.conf.d

Name: PCSC Lite
Description: PC/SC smart card interface
Version: 2.2.3
Libs: -L${libdir} -lpcsclite
Libs.private: -pthread
Cflags: -I${includedir}/PCSC
```

**Comment by emilazy** [comment]
We’re not getting to the linking stage at all, so I don’t think that can be relevant. And the `Cflags` looks fine… but I wonder if `includedir` is causing issues here. Hmm.

**Comment by jnsgruk** [comment]
@SuperSandro2000 any ideas here? My Cmake/pkg-config foo has run out!

**Comment by jnsgruk** [comment]
Hmm, looking through the build output of `pcsclite` if I build it locally, I can see where the `includedir` is being patched and I think it's omitting the prefix?

https://github.com/NixOS/nixpkgs/blob/1bd503aa21f99634e551fa624d8a1a14ccdf5afc/pkgs/build-support/setup-hooks/multiple-outputs.sh#L178

I'll need to do some `sed` research to decode that expression, but I think this might be responsible for omitting `/PCSC` from the end of the `includedir`...

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 any ideas here? My Cmake/pkg-config foo has run out!

No, sorry 😅

**Comment by jnsgruk** [comment]
I think that's a good call. I did manage to make this work by patching the `pcsclite` package to manually fiddle the pkgconfig file, but that doesn't feel like a better route...

```diff
diff --git a/pkgs/tools/security/pcsclite/default.nix b/pkgs/tools/security/pcsclite/default.nix
index eedec6b239fd..82f79202181d 100644
--- a/pkgs/tools/security/pcsclite/default.nix
+++ b/pkgs/tools/security/pcsclite/default.nix
@@ -85,6 +85,10 @@ stdenv.mkDerivation (finalAttrs: {
     moveToOutput bin/pcsc-spy "$dev"
   '';
 
+  postFixup = ''
+    sed -i "/^includedir=/s|=''${dev}/include|=''${dev}/include/PCSC|" $dev/lib/pkgconfig/libpcsclite.pc
+  '';
+
   nativeBuildInputs = [
     meson
     ninja
```

**Pull request merged**